migrator/migrations/m_192_to_m_193_report_config_migrations/migration_impl.go

```diff
@@ -0,0 +1,210 @@
+			return errors.Wrap(err, "failed to scan rows")
+		}
+		// convert to report config proto
+		reportConfigProto, err := updatedSchema.ConvertReportConfigurationToProto(reportConfig)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert %+v to proto", reportConfigProto)
+		}
+		// skip if version=1 since config is migrated
+		// skip if version=2
+		if reportConfigProto.Version != 0 {
+			continue
+		}
+		if reportConfigProto.Version == 0 && reportConfigProto.GetResourceScope() != nil {
+			reportConfigProto.Version = 2
+			// convert report config proto back to gorm model
+			convertedGormConfig, err := updatedSchema.ConvertReportConfigurationFromProto(reportConfigProto)
+			if err != nil {
+				return errors.Wrapf(err, "failed to convert from proto %+v", reportConfigProto)
+			}
+			convertedReportConfigs = append(convertedReportConfigs, convertedGormConfig)
+
+			continue
+		}
+		// create v2 report config from v1
+		newConfig := create_v2_report_config(reportConfigProto)
+
+		// create notifier config for report_configuration_notifier table
+		notifierConfig := create_notifier(reportConfigProto)
+		newConfig.Notifiers = append(newConfig.Notifiers, notifierConfig)
+
+		// add notifier to notifier_configurations_notifiers
+		reportNotifierGorm, err := updatedSchema.ConvertNotifierConfigurationFromProto(notifierConfig, 0, newConfig.GetId())
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert from proto %+v", notifierConfig)
+		}
+		reportNotifiers = append(reportNotifiers, reportNotifierGorm)
+
+		// assign version to 2 to new copy and version to 1 in original so that they are not re-created during migration
+		newConfig.Version = 2
+		reportConfigProto.Version = 1
+		// convert report config proto back to gorm model
+		convertedGormNewConfig, err := updatedSchema.ConvertReportConfigurationFromProto(newConfig)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert from proto %+v", newConfig)
+		}
+		convertedGormReportConfig, err := updatedSchema.ConvertReportConfigurationFromProto(reportConfigProto)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert from proto %+v", reportConfigProto)
+		}
+		convertedReportConfigs = append(convertedReportConfigs, convertedGormNewConfig, convertedGormReportConfig)
+
+		// create report snapshot for last run report job
+		reportSnapshot := create_report_snapshot(reportConfigProto, newConfig)
+		if reportSnapshot != nil {
+			// convert report snapshot to GORM
+			reportSnapshotGORM, err := updatedSchema.ConvertReportSnapshotFromProto(reportSnapshot)
+			if err != nil {
+				return errors.Wrapf(err, "failed to convert from proto %+v", reportSnapshot)
+			}
+			reportSnapshots = append(reportSnapshots, reportSnapshotGORM)
+		}
+		if len(convertedReportConfigs) == batchSize {
+			if err = db.Clauses(clause.OnConflict{UpdateAll: true}).Model(updatedSchema.CreateTableReportConfigurationsStmt.GormModel).Create(&convertedReportConfigs).Error; err != nil {
```

Comment:
Since you are actually updating 3 tables at once, we should probably use a gorm transaction.  (I probably should have thought of that the other day)

https://gorm.io/docs/transactions.html